src/dumbo/src/tcp/endpoint.rs

```diff
@@ -304,64 +305,56 @@ impl Endpoint {
+            StatusCode::BadRequest,
+            Body::new("Invalid request.".to_string()),
+        ),
+        RequestError::HeaderError(err_msg) => {
+            build_response(StatusCode::BadRequest, Body::new(err_msg.to_string()))
+        }
+        RequestError::Overflow => build_response(StatusCode::BadRequest, Body::new(e.to_string())),
+        RequestError::Underflow => build_response(StatusCode::BadRequest, Body::new(e.to_string())),
+    }
+}
+
 /// Parses the request bytes and builds a `micro_http::Response` by the given callback function.
-fn parse_request_bytes(byte_stream: &[u8], callback: fn(Request) -> Response) -> Response {
+fn parse_request_bytes(
+    byte_stream: &[u8],
+    callback: fn(Request, CustomHeaders) -> Response,
+) -> Response {
     let request = Request::try_from(byte_stream);
     match request {
-        Ok(request) => callback(request),
-        Err(e) => match e {
-            RequestError::BodyWithoutPendingRequest => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new(e.to_string()),
-            ),
-            RequestError::HeadersWithoutPendingRequest => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new(e.to_string()),
-            ),
-            RequestError::InvalidHttpVersion(err_msg) => build_response(
-                Version::default(),
-                StatusCode::NotImplemented,
-                Body::new(err_msg.to_string()),
-            ),
-            RequestError::InvalidUri(err_msg) => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new(err_msg.to_string()),
-            ),
-            RequestError::InvalidHttpMethod(err_msg) => build_response(
-                Version::default(),
-                StatusCode::NotImplemented,
-                Body::new(err_msg.to_string()),
-            ),
-            RequestError::InvalidRequest => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new("Invalid request.".to_string()),
-            ),
-            RequestError::HeaderError(err) => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new(err.to_string()),
-            ),
-            RequestError::Overflow => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new(e.to_string()),
-            ),
-            RequestError::Underflow => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new(e.to_string()),
-            ),
+        Ok(request) => match CustomHeaders::try_from(request.headers.custom_map()) {
+            Ok(custom_headers) => callback(request, custom_headers),
+            Err(err) => build_err_response(err),
```

Comment:
This is a miss on my part, MMDS headers parsing should be placed into `/src/mmds/src/lib.rs`. 

I was a little biased because in the initial implementation, I was making use of raw request bytes to parse custom headers. This is why this operation needed to be inside `parse_request_bytes()` function (`dumbo` crate), last place in the stack where the raw bytes where still accessible. 

However, after I improved the implementation, I omitted that custom headers parsing no longer depend on request bytes. I will move the MMDS specific logic to `mmds` crate.